processing/src/main/java/org/apache/druid/query/QueryContexts.java

```diff
@@ -174,6 +179,35 @@
     }
   }
 
+  private static int checkPositive(String propertyName, int val)
+  {
+    Preconditions.checkArgument(
+        val > 0,
+        "%s should be positive, but [%s]",
+        propertyName,
+        val
+    );
+    return val;
+  }
+
+  /**
+   * Return the configured number of combine threads if any. Others {@link #NO_PARALLEL_COMBINE_THREADS}.
+   */
+  public static <T> int getNumBrokerParallelCombineThreads(Query<T> query)
+  {
+    return parseInt(query, NUM_BROKER_PARALLEL_COMBINE_THREADS, NO_PARALLEL_COMBINE_THREADS);
```

Comment:
From this line it seemed to me that the default is not parallelizing by default.

Reply:
Yes, it's not enabled by default because I'm not sure how this affects to the overall performance of the cluster if it's enabled for all queries. Also, enabling parallel combine might not be good if the size of intermediate aggregates is not large enough which is the common Druid use case. I added this to doc.

Reply:
> Is it possible to manage processing resources so that there is confidence that even if parallel combine is enabled for all queries, it doesn't make things much worse (but parallelization overhead is acceptable, e. g. within 10% of regression in the most unfavourable cases)?

Here is how it currently works.

- Brokers use the blockingPool to manage processing buffers.
- If `NUM_BROKER_PARALLEL_COMBINE_THREADS` = -1 for a query, the broker reserves all current available resources to the query. Once the query computes the actual required number of buffer pools among the available ones, it immediately releases them (https://github.com/apache/incubator-druid/pull/6629/files#diff-dc9ea12c892e41bbd957c77b43810ddeR103). If there are not enough buffers available, the query is simply run without parallel combine.
- if `NUM_BROKER_PARALLEL_COMBINE_THREADS` is a positive number _n_, the broker reserves _n_ buffers. If there's not enough buffers, the query waits for _n_ buffers to be available.

So, for your question, if `NUM_BROKER_PARALLEL_COMBINE_THREADS` is set to -1 for all queries, all queries can be run at least without waiting for buffers which I think is the bottom line.

> Is it possible to estimate the size of intermediate aggregates and enable parallel combine automatically when it begins to make sense?

Well, we can do some rough optimization based on the estimated input size and the cardinality of grouping keys, but not sure how good it will be.

In the future, I think it's possible to do advanced optimizations for queries which need blocking aggregation in historicals, like groupBy. Since, in blocking aggregation, historicals are able to know the exact size and number of intermediate aggregates they will send to the broker, they can send those metadata before sending the actual aggregates, so that the broker can make some decisions about run-time optimization like parallel merge. Does this make sense?

Reply:
How does ForkJoinPool / parallel Streams facility approach this problem? Is it applicable to parallel combine in brokers?

> Well, we can do some rough optimization based on the estimated input size and the cardinality of grouping keys, but not sure how good it will be.

I think the advantage that we have and parallel streams don't is that we can safely assume that all pieces of data to be merged take approximately the same time. So we could extrapolate, based on processing of some small part of data.

Reply:
I think there're at least two algorithms applicable with ForkJoinPool.

1. Same algorithm with the one implemented in this PR. Combining sequences is distributed to multiple ForkJoinTasks, and these tasks are pipelined. So, each ForkJoinTask combines input sequences and stores outputs in its blocking queue. 
2. Splitting each sequence into small chunks and combines corresponding chunks from different sequences in parallel. Each ForkJoinTask combines a set of chunks of sequences. I think this is more natural way to use ForkJoinPool because we need to fork tasks to combines small chunks and then join to compute the final aggregates. However, this requires to materialize intermediate aggregates which should be avoided.

So, I'm not sure what the benefit of using ForkJoinPool is here. At least we don't have to add a new configuration for ForkJoinPool by using the existing processing threads.

> I think the advantage that we have and parallel streams don't is that we can safely assume that all pieces of data to be merged take approximately the same time. 

What kind of algorithms are you thinking? Would you elaborate more on the assumption?

Reply:
What I want is that there is no `NUM_BROKER_PARALLEL_COMBINE_THREADS` config, or at least it defaults to "parallelize ergonomically".

As far as I can tell the main struggle is that is a very large combine task arrives, we want to give it as much parallelism as possible, but not to block concurrently arriving small tasks.

It think it could be accomplished by:
 1) start processing in a `ForkJoinTask`, checking the total processing time (using `System.nanoTime()`) e. g. after each 1000 elements processed.
 2) If after some point the total processing time exceeds 10ms, we divide the number of total elements to combine by the number of currently processed, and get the number of chunks, processing of each of them should take approximately 10ms (we should be pretty confident about it, because processing of each element should take deterministic time)
 3) ForkJoinTask reforks itself into another task (preserving and using the same combine state), that processes the number of elements equal to the chunk size, and then reforks again, and so on until the elements are exhausted. All using the same state.

It means that processing is still essentially sequential, just with artificial "forking points" each ~10 ms. Now to ensure that small combine tasks are not blocked, probably we just need to use a ForkJoinPool in [`asyncMode`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html#ForkJoinPool-int-java.util.concurrent.ForkJoinPool.ForkJoinWorkerThreadFactory-java.lang.Thread.UncaughtExceptionHandler-boolean-).

Reply:
To ensure that all sequences are still processed timely, and efficiently due to  good locality (cache utilization, etc) probably the number of sequences concurrently processed in such ForkJoinPool should be limited to about nThreads * 1.5, where nThreads is the number of threads in the pool.

To ensure that all queries are processed timely too, the sequences could be feeded to the ForkJoinPool only from a fixed set of queries, in a round-robin fashion. Excess queries are queued.

Reply:
@leventov would you explain how the algorithm you described makes the combine faster?

Reply:
I'm not sure if it makes the alrogithm faster or not, but it allows to enable parallelism by default, that is critically important IMO.

Reply:
@leventov I'm still not sure what your idea is to achieve. Currently, a broker can combine multiple queries at the same time without blocking each other, although the number of concurrent queries is limited by the number of its http threads. I'm not sure how your idea is better than now.

Do you just want to avoid the manual setting for `NUM_BROKER_PARALLEL_COMBINE_THREADS`? My intention was we need a way to explicitly specify the number of threads for combine because automatic optimization might be wrong or not enough. (This is especially true in this PR because the only available option for auto optimization is to set `NUM_BROKER_PARALLEL_COMBINE_THREADS` to -1.) However, if they really want to use some particular number of threads, there should be enough available threads at the query time. Otherwise, the query will be blocked and failed in the end.

What do you think about this? And do you think it makes sense to add a better optimization mechanism later?

Reply:
It's essentially impossible to set `NUM_BROKER_PARALLEL_COMBINE_THREADS` "right" even if it is done on query per query basis, that will nobody ever do. This moment broker could be idle, the next moment a lot of queries will arrive and setting `NUM_BROKER_PARALLEL_COMBINE_THREADS` will only make things worse.

> because automatic optimization might be wrong or not enough

My intention to employ techniques from ForkJoinPool. When you use  `parallelStream()`, it basically could not parallelize wrong or not enough. It may not be 100% optimal because of coarse partitioning granularity (You could only split any piece of work in two parts, not three, for example), but it is always within 10-20% from optimum. And this actually should not be so for combine on brokers.

The only way to really abuse `parallelStream()` is to pass it something that could be completed in less than 10ms. But it could be safely avoided in parallel combine too, as I shown above.

Reply:
@leventov sorry, maybe I wasn't clear. What I don't understand is how your idea makes the combine faster than at least the current implementation in master. 

I talked about this with @jon-wei and maybe the following is what you're thinking.

You may think of the below architecture for parallel combine. `H{n}` and `F{n}` represent a HTTP thread and a Fork Join Task, respectively. `S{n}` is a sequence.

```
 H1 H2 H3 H4 H5 (HTTP thread pool)
  \  \ | /  /
  +----------+
  | F1 F2 F3 | (Fork Join Pool)
  +----------+
 /  /  / \  \  \
S1 S2 S3 S4 S5 S6
```

The fork join tasks can combine any sequences for any queries, but the sequences to combine together should be for the same query. Also, each fork join task computes for a limited time like 10ms as you said. Once they complete their task, they return the result to the proper HTTP thread. The HTTP threads combine the results from fork join tasks again to get the final result. For a query, since the combine is hierarchically computed in an HTTP thread and fork join tasks, it can be done in parallel and be faster than the current implementation in master. If this is what you think, I think it's a good idea.

Reply:
I didn't understand what do you mean, but we could do the initial "sampling" merge of just two sequences, and then, based on the timing and the number of idle threads in FJP (`getRunningThreadCount()`, `getActiveThreadCount()`, etc.) decide on the degree of merge, similarly to what you already do in `ParallelMergeCombineSequence`. So it's the same hierarchical merge, just done in FJP with "yields" every 10ms, so that small queries that arrived in parallel with the humongous query have a chance to intercept and don't have to wait long until that big query combine is completed, because all processing threads are busy.